replica/database.cc

```diff
@@ -2359,12 +2359,21 @@ future<> database::flush_keyspace_on_all_shards(sharded<database>& sharded_db, s
     });
 }
 
-future<> database::flush_tables_on_all_shards(sharded<database>& sharded_db, std::vector<table_id> table_ids) {
-    co_await sharded_db.invoke_on_all([] (replica::database& db) {
-        return db._commitlog->force_new_active_segment();
-    });
-    co_await coroutine::parallel_for_each(std::move(table_ids), [&] (auto& table_id) {
-        return flush_table_on_all_shards(sharded_db, table_id);
+future<> database::flush_dirty_tables_on_all_shards(sharded<database>& sharded_db, std::vector<table_id> table_ids) {
```

Comment:
As I explained in the earlier patch, I think that passing a table_id vector to this function is wrong.
The list of dirty tables needs to be constructed bottom-up rather than top-down (and filtered).
We need to go over all non-active commitlog segments and harvest the tables that refer to them into a set (can be unrdered_set), and flush those.

It would be nice to refine that even further for tablets by collecting not just `table_id`s, but also token ranges, i.e. the dirty compaction_groups/tablets and flush only those, but that can be done as a follow up.

Reply:
@bhalevy is this change supposed to address the tablet case? IIUC, it's tracked by #16752. or do you want me to handle tablet in here as well?

Reply:
No it's

Reply:
> Commitlog may contain pieces of unrelated tables that are not part of the scope of this force_keyspace_cleanup.

Oh, I see. I had no idea that the user-visible cleanup API is per-keyspace. That doesn't really make sense.

Reply:
despite that we should not *only* flush the given tables by nodetool, i think, we can still dispense with the "force new active commitlog segment" step, if active commitlog has no mutations touching these tables.

for instance, `nodetool` is instructed to cleanup keyspace "ks2", but the active commitlog segments only contains the mutations to "ks1". without forcing new active segment, we can still cleanup "ks2" on current node by flushing memtables and sstables belonging to this keyspace.